npm/design-system/src/components/FileExplorer/FileExplorer.tsx

```diff
@@ -0,0 +1,177 @@
+   * only from a user's point of view.
+   * For this reason we save the open state as part of the UI component. The easiest
+   * way to do this is a key/value pair, mapping the absolute path of a directory to a boolean
+   *
+   * {
+   *   'foo': true,
+   *   'foo/bar': true
+   *   'foo/bar/qux': false
+   * }
+   *
+   * Every directory is set to open by default. When you add a new directory
+   * or file via your file system (eg mkdir foo/bar && touch foo/bar/hello.js) it will be added
+   * without losing the current state of open/closed directories.
+   */
+  const [openFolders, setOpenFolders] = React.useState<Record<string, boolean>>({})
+
+  React.useEffect(() => {
+    function walk (nodes: TreeNode[]) {
+      for (const node of nodes) {
+        if (node.type === 'folder') {
+          if (!(node.absolute in openFolders)) {
+            setOpenFolders({ ...openFolders, [node.absolute]: true })
+          }
+
+          walk(node.files)
+        }
+      }
+    }
+
+    walk(props.files)
+  }, [props.files, openFolders])
+
+  const setSelectedFile = (absolute: string) => {
+    setOpenFolders({ ...openFolders, [absolute]: !openFolders[absolute] })
+  }
+
+  return (
+    <nav className={cs(props.className, props.cssModule.nav)}>
+      <FileTree
+        {...props}
+        setSelectedFile={setSelectedFile}
+        openFolders={openFolders}
+        depth={0}
+      />
+    </nav>
+  )
+}
+
+export const FileTree: React.FC<FileTreeProps> = (props) => {
+  // Negative margins let the <a> tag take full width (a11y)
+  // while the <li> tag with text content can be positioned relatively
+  // This gives us HTML + cssModule-only highlight and click handling
+  const inlineStyles = {
+    a: {
+      marginLeft: `calc(-20px * ${props.depth})`,
+      width: `calc(100% + (20px * ${props.depth}))`,
+    },
+    li: {
+      marginLeft: `calc(20px * ${props.depth})`,
+    },
+  }
+
+  const fileTree = (item: TreeNode) => {
+    if (item.type !== 'folder') {
+      return
+    }
+
+    return (
+      <FileTree
+        fileComponent={props.fileComponent}
+        folderComponent={props.folderComponent}
+        openFolders={props.openFolders}
+        setSelectedFile={props.setSelectedFile}
+        onFileClick={props.onFileClick}
+        selectedFile={props.selectedFile}
+        depth={props.depth + 1}
+        cssModule={props.cssModule}
+        files={props.openFolders[item.absolute] ? item.files : []}
+      />
+    )
+  }
+
+  const renderFolder = (item: FolderNode) => {
+    return (
+      <props.folderComponent
+        depth={props.depth}
+        item={item}
+        isOpen={props.openFolders[item.absolute]}
+        onClick={() => props.setSelectedFile(item.absolute)}
+      />
+    )
+  }
+
+  const renderFile = (item: FileNode) => {
+    return (
+      <props.fileComponent
+        depth={props.depth}
+        item={item}
+        onClick={props.onFileClick}
+      />
+    )
+  }
```

Comment:
This is stuff you really want to prevent in `render`, as it adds additional cost due to the dynamic nature of your JSX. It would be better if these were statically defined components, and those components received all of the props necessary.

This particular situation isn't as bad because you're only adding new JSX nodes with the functions, however, if you were using these to create callbacks, there would be a real performance impact.

Reply:
Hm, my goal here was to let the user provide their own implementation for the file/folders. Is there some other pattern to accomplish this?

Reply:
To make them entirely static, you can make your props take either `JSX.Element`, or an _actual_ React component (`React.Component<YourPropType>`). With `JSX.Element`, you would pass just a normal JSX component invocation to your prop. This won't work since you want to bind to your item (well, you can make it work, but I won't go into that).

Instead, you could define your prop interface, then pass a React component that receives all of those props and does something with it. In that scenario you'd be passing the actual React function (or class) instance.

Finally, if you do want to do "render props" (which is essentially what you've done), you probably want to wrap the functions in `useCallback`.

I want to emphasize that with how you have written this, I don't think any change is necessary, and the first two suggestions would only bring extremely minor performance fixes. I was more just commenting for learning purposes.

Reply:
+1 for renderFile and other render props to be passed directly.